support namespace indexer for namespaced resources like pods

ahutsunshine opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
In our cluster, the number of some resources like pods are very large(10w+). When listing a namespace resources like pods from apiserver cache, the apiserver process will firstly get all the resources in all the namespaces and then filter the resources by namespace predicate even it's a very small request. It's heavy for filtering namespaced resource even the resource number in the namespace is very small with many client requests. It consumes lots of cpu & memory that's not expected.

#### Which issue(s) this PR fixes:
feature support and discussion: https://github.com/kubernetes/kubernetes/issues/120778

```release-note
NONE
```

HirazawaUi commented:
If we need to add a feature gate, [kep](https://github.com/kubernetes/enhancements) is a good place to start, but before we do that, I want to know if we really need it

ahutsunshine commented:
@HirazawaUi We do a benchmark and see perfect performance to improve latency&cpu&memory. I proposal this because listing the pods across various namespaces significantly impacts CPU and memory resources and causes the incidents in our production. You can check more https://github.com/kubernetes/kubernetes/issues/120778 and we can discuss the feature/enhancements.

bart0sh commented:
Thank you for your PR. Please [sign the CLA](https://github.com/kubernetes/community/tree/master/contributors/guide/contributor-cheatsheet#signing-the-cla) to proceed further, thanks.

halfcrazy commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -746,7 +747,34 @@ func (c *Cacher) listItems(ctx context.Context, listRV uint64, key string, pred
+	}
+
+	if pred.Field.Empty() {
+		// add default metadata.namespace field selector if not set for pod list
```
Nit: not only pod kind

halfcrazy commented:
The namespace indexer only adds to some in-tree resources. Could this work for custom resources as well?

jiahuif commented:
Also, pinging for CLA.
/triage accepted

ahutsunshine commented:
I refactor the PR as we agreed in https://github.com/kubernetes/kubernetes/issues/120778#issuecomment-1829322398 that  currently it only supports pod namespace indexer. If working and showing perfect performance in some cases, then we will go ahead to support other resources.  @halfcrazy @caseydavenport @deads2k @jiahuif

ahutsunshine commented:
/easycla

ahutsunshine commented:
/easycla

wojtek-t commented:
Sometimes closing and reopening PR may help - let me try.

wojtek-t commented on pkg/features/kube_features.go:
```diff
@@ -1281,4 +1287,6 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	// ...
 	HPAScaleToZero: {Default: false, PreRelease: featuregate.Alpha},
+
+	NamespaceIndex: {Default: false, PreRelease: featuregate.Alpha},
```
We can probably go directly to beta - we won't get anything from using alpha disabled by default.

wojtek-t commented on pkg/features/kube_features.go:
```diff
@@ -975,6 +975,12 @@ const (
+	// alpha: v1.29
+	//
+	// Allows namespace indexer for namespace scope resources in apiserver cache to accelerate list operations.
+	NamespaceIndex featuregate.Feature = "NamespaceIndex"
```
Let's maybe call it StorageNamespaceIndex

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -746,7 +747,33 @@ func (c *Cacher) listItems(ctx context.Context, listRV uint64, key string, pred
+
+func matcherIndex(ctx context.Context, pred storage.SelectionPredicate) []storage.MatchValue {
+	re, _ := request.RequestInfoFrom(ctx)
+	if re == nil || !re.IsResourceRequest || re.Subresource != "" || !whitelistResource(re.Resource) {
```
We shouldn't hardcode resources here - instead you should do that purely relying on whethever `IndexFields` contains `metadata.namespace` inside.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -746,7 +747,33 @@ func (c *Cacher) listItems(ctx context.Context, listRV uint64, key string, pred
+	// 2.1 pred.Field will not have "metadata.namespace" field if the request is like /api/v1/pods
+	// 2.2 pred.Field will have "metadata.namespace" field if the request is like /api/v1/pods?fieldSelector=metadata.namespace=ns
+	return pred.MatcherIndex()
+}
```
I don't think all this logic belongs here. Instead I would rather switch `MatcherIndex()` function to take a context as a parameter and use it there [without modifying pred itself].

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:
```diff
@@ -746,7 +747,33 @@ func (c *Cacher) listItems(ctx context.Context, listRV uint64, key string, pred
+
+// whitelistResource only allows pod to pass that represents only listing pods will enable namespace indexer
+func whitelistResource(resource string) bool {
+	return strings.ToLower(resource) == "pods"
```
Also - we definitely need some unit test for showing that the index is used.

ahutsunshine replied:
Added the ut

wojtek-t commented:
/easycla

wojtek-t commented:
> https://github.com/kubernetes/kubernetes/pull/121906#issuecomment-1831298776

Can you please look into CLA again - from that screenshot there is "SIGN CLA" button there [it only says "document ready for signature"]

ahutsunshine committed:
support pod namespace indexer

fix comments

optimize code

small optimization for the namespace scope check

ahutsunshine commented:
> > [#121906 (comment)](https://github.com/kubernetes/kubernetes/pull/121906#issuecomment-1831298776)
> 
> Can you please look into CLA again - from that screenshot there is "SIGN CLA" button there [it only says "document ready for signature"]

Fixed. I use company's user name. Changed to my personal account.

wojtek-t commented:
/ok-to-test

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/selection_predicate.go:
```diff
@@ -145,9 +149,44 @@ func (s *SelectionPredicate) Empty() bool {
 // of objects that return <value> for a given index, a pair (<index name>, <value>)
 // wil be returned.
-func (s *SelectionPredicate) MatcherIndex() []MatchValue {
+func (s *SelectionPredicate) MatcherIndex(ctx context.Context) []MatchValue {
```
I'm not a fan of the code below - I think sth like that would be easier:
```
for _, field := range s.IndexFields {
	if value, ok := s.Field.RequiresExactMatch(field); ok {
		result = append(result, MatchValue{IndexName: FieldIndex(field), Value: value})
	} else if field == "metadata.namespace" {
		if namespace, isNamespaceScope := isNamespaceScopedRequest(ctx); isNamespaceScope {
			result = append(result, MatchValue{IndexName: FieldIndex(field), Value: namespace})
		}
	}
	...
```

namespace.name is somewhat special anyway, so I think that making this explicit would make it much easier to follow

ahutsunshine replied:
Updated.

wojtek-t replied:
I very consciously wrote what is above and what you did is not that.
There are couple reasons I did what I did being:
(a) the other pod index is `spec.nodeName` and in many cases that is much more limiting index.
  So if such fieldSelector is specified we actually want to prioritize for it [and that is set as the first one in `IndexFields`
(b) let's not even try to look at request info if we can use `nodeName` index
(c) if FieldSelector itself is setting `metadata.namespace` let's prefer this one

ahutsunshine replied:
Thanks for your detail explanation.  I consider your code, but there is one case it doesn't handle correctly. i.e. `/api/v1/namespaces/default/pods?fieldSelector=metadata.namespace=kube-system`. The above code will try to get pods from `kube-system` namespace and then filter data by the namespace `kube-system` where it's not correct.
For the `spec.nodeName`, you're right and I miss the situation.  I update the codes.

ahutsunshine commented on staging/src/k8s.io/apiserver/pkg/storage/selection_predicate.go:
```diff
@@ -145,9 +149,44 @@ func (s *SelectionPredicate) Empty() bool {
+
+// whitelistResource only allows pod to pass that represents only listing pods will enable namespace indexer
+func whitelistResource(resource string) bool {
+	return strings.EqualFold(resource, "pods")
```
Actually I don't get the point. If we don't hardcode to check the resource, the following logic may take `metadata.namespace` index for other resources like configmaps that doesn't make sense and will get wrong data. Can you point out the code more clear?  Thanks @halfcrazy .

ahutsunshine commented:
@wojtek-t @halfcrazy pls go ahead to review

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/selection_predicate.go:
```diff
@@ -160,6 +168,14 @@ func (s *SelectionPredicate) MatcherIndex() []MatchValue {
 
+func isNamespaceScopedRequest(ctx context.Context) (string, bool) {
+	re, _ := request.RequestInfoFrom(ctx)
+	if re == nil {
```
nit:
```
if re == nil || len(re.Namespace) == 0 {
  return "", false
}
return re.Namespace, true
```

ahutsunshine replied:
Sure. Updated.

wojtek-t reviewed (commented):
Just one minor nit - other than that LGTM.

ahutsunshine commented:
> LGTM - please squash the commits and I will approve.

Really appreciate the review and recommendations.  Squash the commits and commit.

ahutsunshine commented:
@halfcrazy @caesarxuchao @deads2k Feel free to review the PR if necessary. @wojtek-t gives LGTM and approval.

ahutsunshine commented:
/milestone v1.29

wojtek-t commented:
No - we're very deep in code freeze (right before release) and this shouldn't go to 1.29.
This will merge once the branch opens after release.

MadhavJivrajani commented:
Considering this adds in a feature gate, its probably worth adding a release note since we aren't going down the KEP route.

ahutsunshine commented on pkg/features/kube_features.go:
```diff
@@ -975,6 +975,12 @@ const (
 	UserNamespacesPodSecurityStandards featuregate.Feature = "UserNamespacesPodSecurityStandards"
+
+	// owner: @ahutsunshine
+	// beta: v1.29
```
@MadhavJivrajani Thanks for your reminding. Yeah, it's a mistake. Shall we file a PR to fix it?

MadhavJivrajani replied:
Yes please do! Thank you!

ahutsunshine replied:
PR: https://github.com/kubernetes/kubernetes/pull/122732
